File: language/benchmarks/tokens/cases/concurrent-region/concurrent-region.py
import asyncio

ConcurrentOutcome = dict[str, int | str]
Result = tuple[bool, int | str]


async def sleep_ms(ms: int) -> None:
    await asyncio.sleep(ms / 1000)


async def main() -> list[ConcurrentOutcome]:
    return await asyncio.gather(*(run_process(value) for value in [1, 2, 3]))


async def process(value: int) -> Result:
    await sleep_ms(0)
    return (True, value)


async def run_process(value: int) -> ConcurrentOutcome:
    ok, payload = await process(value)
    return {"kind": "success", "value": payload} if ok else {"kind": "failure", "message": payload}

File: language/benchmarks/tokens/cases/concurrent-region/concurrent-region.ts
type ConcurrentOutcome<T,E>=
  | {kind:"success",value:T}
  | {kind:"failure",message:E}
  | {kind:"aborted"};

type Result<T,E>={ok:true,value:T}|{ok:false,error:E};

function sleepMs(ms:number):Promise<void>{
  return new Promise((resolve)=>setTimeout(resolve,ms));
}

async function main():Promise<ConcurrentOutcome<number,string>[]>{
  return Promise.all([1,2,3].map(runProcess));
}

async function process(value:number):Promise<Result<number,string>>{
  await sleepMs(0);
  return {ok:true,value};
}

async function runProcess(value:number):Promise<ConcurrentOutcome<number,string>>{
  const result=await process(value);
  return result.ok ? {kind:"success",value:result.value} : {kind:"failure",message:result.error};
}
